diff --git a/src/gallium/drivers/r600/r600_pipe.c b/src/gallium/drivers/r600/r600_pipe.c
index 435aa77..c2b1fe1 100644
--- a/src/gallium/drivers/r600/r600_pipe.c
+++ b/src/gallium/drivers/r600/r600_pipe.c
@@ -904,6 +904,7 @@ struct pipe_screen *r600_screen_create(struct radeon_winsys *ws)
        case R600:
        case EVERGREEN:
                rscreen->has_streamout = rscreen->info.drm_minor >= 13;
+                fprintf(stderr, "drm_minor : %d\n", rscreen->info.drm_minor);
                break;
        case R700:
                rscreen->has_streamout = rscreen->info.drm_minor >= 17;

