diff --git a/configure.ac b/configure.ac
index af5f4c6..9bda55c 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1344,7 +1344,7 @@ fi
 
 AC_ARG_WITH([clang-libdir],
    [AS_HELP_STRING([--with-clang-libdir],
-         [Path to Clang libraries @<:@default=llvm-config --libdir@:>@])],
+         [Path to Clang libraries @<:@default=llvm-config-3.5 --libdir@:>@])],
    [CLANG_LIBDIR="$withval"],
    [CLANG_LIBDIR=''])
 
@@ -1574,9 +1574,9 @@ if test "x$enable_gallium_llvm" = xauto; then
 fi
 if test "x$enable_gallium_llvm" = xyes; then
     if test -n "$llvm_prefix"; then
-        AC_PATH_TOOL([LLVM_CONFIG], [llvm-config], [no], ["$llvm_prefix/bin"])
+        AC_PATH_TOOL([LLVM_CONFIG], [llvm-config-3.5], [no], ["$llvm_prefix/bin"])
     else
-        AC_PATH_TOOL([LLVM_CONFIG], [llvm-config], [no])
+        AC_PATH_TOOL([LLVM_CONFIG], [llvm-config-3.5], [no])
     fi
 
     if test "x$LLVM_CONFIG" != xno; then
