diff --git a/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp b/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp
index 1e5adb7..7b56e1e 100644
--- a/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp
+++ b/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp
@@ -70,7 +70,11 @@
 #if HAVE_LLVM >= 0x0303
 #include <llvm/IR/IRBuilder.h>
 #include <llvm/IR/Module.h>
+#endif
+#if HAVE_LLVM > 0x0303
 #include <llvm/Support/CBindingWrapping.h>
+#else
+#include <llvm/Wrap.h>
 #endif
 
 #include "pipe/p_config.h"
