diff --git a/src/gallium/drivers/r600/r600_buffer.c b/src/gallium/drivers/r600/r600_buffer.c
index 6df0d91..82630af 100644
--- a/src/gallium/drivers/r600/r600_buffer.c
+++ b/src/gallium/drivers/r600/r600_buffer.c
@@ -137,6 +137,7 @@ static void *r600_buffer_transfer_map(struct pipe_context *ctx,
 			r600_set_constants_dirty_if_bound(rctx, rbuffer);
 		}
 	}
+#if 0 /* this is broken (see Bug 53130) */
 	else if ((usage & PIPE_TRANSFER_DISCARD_RANGE) &&
 		 !(usage & PIPE_TRANSFER_UNSYNCHRONIZED) &&
 		 rctx->screen->has_streamout &&
@@ -161,6 +162,7 @@ static void *r600_buffer_transfer_map(struct pipe_context *ctx,
 			}
 		}
 	}
+#endif
 
 	/* mmap and synchronize with rings */
 	data = r600_buffer_mmap_sync_with_rings(rctx, rbuffer, usage);
