diff --git a/configure.ac b/configure.ac
index fc7d372..82e47c8 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1981,7 +1981,13 @@ dnl in LLVM_LIBS.
 
 if test "x$MESA_LLVM" != x0; then
 
-    LLVM_LIBS="`$LLVM_CONFIG --libs ${LLVM_COMPONENTS}`"
+    if test $LLVM_VERSION_MAJOR -eq 3 -a $LLVM_VERSION_MINOR -ge 5 ; then
+        LLVM_LIBS="`$LLVM_CONFIG --libs ${LLVM_COMPONENTS}`"
+        LLVM_LIBS+=" `$LLVM_CONFIG --system-libs ${LLVM_COMPONENTS}`"
+        dnl Because my llvm-config adds a new line...
+    else
+        LLVM_LIBS="`$LLVM_CONFIG --libs ${LLVM_COMPONENTS}`"
+    fi
 
     if test "x$enable_llvm_shared_libs" = xyes; then
         dnl We can't use $LLVM_VERSION because it has 'svn' stripped out,
